[AutoDiff upstream] Add `@differentiable` declaration attribute type-checking.

dan-zheng opened the pull request:
The `@differentiable` attribute marks a function as differentiable.

Example:
```swift
@differentiable(wrt: x, jvp: derivativeFoo where T: Differentiable)
func id<T>(_ x: T) -> T { x }
```

The `@differentiable` attribute has an optional `wrt:` clause specifying the
parameters that are differentiated "with respect to", i.e. the differentiability
parameters. The differentiability parameters must conform to the
`Differentiable` protocol.

If the `wrt:` clause is unspecified, the differentiability parameters are
currently inferred to be all parameters that conform to `Differentiable`.

The `@differentiable` attribute also has optional `jvp:` and `vjp:` labels
for registering derivative functions. These labels are deprecated in favor of
the `@derivative` attribute and will be removed soon.

The `@differentiable` attribute also has an optional `where` clause, specifying
extra differentiability requirements for generic functions.

The `@differentiable` attribute is gated by the
`-enable-experimental-differentiable-programming` flag.

---

Type-checking rules (summary):
- `@differentiable` attribute must be declared on a function-like "original"
  declaration: `func`, `init`, `subscript`, `var` (computed properties only).
- Parsed differentiability parameters must be valid (if they exist).
- Parsed `where` clause must be valid (if it exists).
- Differentiability parameters must all conform to `Differentiable`.
- Original result must all conform to `Differentiable`.
- If JVP/VJP functions are specified, they must match the expected type.
  - `@differentiable(jvp:vjp:)` for derivative registration is deprecated in
    favor of `@derivative` attribute, and will be removed soon.
- Duplicate `@differentiable` attributes with the same differentiability
  parameters are invalid.
- For protocol requirements and class members with `@differentiable` attribute,
  conforming types and subclasses must have the same `@differentiable` attribute
  (or one with a superset of differentiability parameter indices) on
  implementing/overriding declarations.

These rules are consistent with the [differentiable programming manifesto](https://github.com/apple/swift/blob/master/docs/DifferentiableProgramming.md#the-differentiable-declaration-attribute-1).
The main proposed rules are implemented.

Code changes:
- Add `DifferentiableAttributeTypeCheckRequest`.
  - Currently, the request returns differentiability parameter indices,
    while also resolving `JVPFunction`, `VJPFunction`, and
    `DerivativeGenericSignature` and mutating them in-place in
    `DifferentiableAttr`. This works fine for now.
- Add "is type-checked" bit to `DifferentiableAttr`.
  - Alternatively, I tried changing `CacheKind::SeparatelyCached` to
    `CacheKind::Cached`, but it did not seem to work: `@differentiable`
    attributes in non-primary-files were left unchecked.

Upstream disorganized tests as-is from `tensorflow` branch.

Resolves TF-828: upstream `@differentiable` attribute type-checking.

dan-zheng committed:
[AutoDiff upstream] Add `@differentiable` attribute type-checking.

Type-checking rules (summary):
- `@differentiable` attribute must be declared on a function-like "original"
  declaration: `func`, `init`, `subscript`, `var` (computed properties only).
- Parsed differentiability parameters must be valid (if they exist).
- Parsed `where` clause must be valid (if it exists).
- Differentiability parameters must all conform to `Differentiable`.
- Original result must all conform to `Differentiable`.
- If JVP/VJP functions are specified, they must match the expected type.
  - `@differentiable(jvp:vjp:)` for derivative registration is deprecated in
    favor of `@derivative` attribute, and will be removed soon.
- Duplicate `@differentiable` attributes with the same differentiability
  parameters are invalid.
- For protocol requirements and class members with `@differentiable` attribute,
  conforming types and subclasses must have the same `@differentiable` attribute
  (or one with a superset of differentiability parameter indices) on
  implementing/overriding declarations.

Code changes:
- Add `DifferentiableAttributeTypeCheckRequest`.
  - Currently, the request returns differentiability parameter indices,
    while also resolving `JVPFunction`, `VJPFunction`, and
    `DerivativeGenericSignature` and mutating them in-place in
    `DifferentiableAttr`. This works fine for now.
- Add "is type-checked" bit to `DifferentiableAttr`.
  - Alternatively, I tried changing `CacheKind::SeparatelyCached` to
    `CacheKind::Cached`, but it did not seem to work: `@differentiable`
    attributes in non-primary-files were not type-checked for some reason.

Upstream disorganized tests as-is from `tensorflow` branch.

Resolves TF-828.

dan-zheng commented:
Let's see if tests pass.
@swift-ci Please smoke test

dan-zheng commented on include/swift/AST/Decl.h:
```diff
@@ -6060,6 +6074,12 @@ class AbstractFunctionDecl : public GenericContext, public ValueDecl {
   bool hasDynamicSelfResult() const;
 
+  /// Get all derivative function configurations.
+  ArrayRef<AutoDiffConfig> getDerivativeFunctionConfigurations();
```
Note: `AbstractFunctionDecl::getDerivativeFunctionConfigurations` is used for derivative function lookup (SILOptimizer stuff) and actually isn't relevant for `@differentiable` attribute type-checking. For now, `ASTContext::DifferentiableAttrs` is used to check duplicate `@differentiable` attributes.

I realized this too late - TF-1100 tracks the rest of upstreaming `AbstractFunctionDecl::getDerivativeFunctionConfigurations`.

dan-zheng replied:
Actually, `AbstractFunctionDecl::getDerivativeFunctionConfigurations` is used in `TypeCheckProtocol.cpp` for verifying that protocol requirement `@differentiable` attributes are satisfied by witnesses.

dan-zheng commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -530,7 +530,114 @@ swift::matchWitness(
 
   // Now finalize the match.
-  return finalize(anyRenaming, optionalAdjustments);
+  auto result = finalize(anyRenaming, optionalAdjustments);
```
Note: I anticipate that maybe some code owners here would care strongly about preserving `return finalize(anyRenaming, optionalAdjustments);`. It doesn't matter too much to me, so I upstreamed code from `tensorflow` branch as in.

Aside: this logic for diagnosing unmet `@differentiable` attribute requirement doesn't work when associated type inference is involved in witness checking. TF-1014 tracks that issue.

dan-zheng replied:
> Note: I anticipate that some code owners might care strongly about preserving `return finalize(anyRenaming, optionalAdjustments)` here.

Slightly mitigating the above point: I refactored `swift::matchWitness` logic for checking `@differentiable` attributes into a `matchWitnessDifferentiableAttr` helper. So the code changes to `swift::matchWitness` are minimal.

rxwei reviewed (approved):
LGTM. Please clarify on the PR title that this is "Adding `@differentiable` **declaration** attribute type-checking" to avoid any confusion with the `@differentiable` function type attribute.

marcrasi commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -530,7 +530,114 @@ swift::matchWitness(
 
   // Now finalize the match.
-  return finalize(anyRenaming, optionalAdjustments);
+  auto result = finalize(anyRenaming, optionalAdjustments);
```
What do you think about adding this separately, and just forbidding @diffble requirements on master for now? It would allow you to make this PR smaller (also you'll be able to exclude the whole derivative function configuration thing), and would let the code owners review this as an independent patch where we give more context about how it works and what it enables.

dan-zheng replied:
I'm open to splitting this PR, but I won't do so preemptively. I plan to cc a code owner soon - we can split if requested.

marcrasi commented on lib/Sema/TypeCheckDeclOverride.cpp:
```diff
@@ -595,6 +595,106 @@ static bool parameterTypesMatch(const ValueDecl *derivedDecl,
 
+/// Returns true if `derivedDecl` has a `@differentiable` attribute that
+/// overrides one from `baseDecl`.
+static bool hasOverridingDifferentiableAttribute(ValueDecl *derivedDecl,
```
Similar to my comment below about protocol requirements, it might be good to exclude this from this PR.

dan-zheng replied:
I'm open to splitting this PR, but to save time, I won't do so preemptively. I'll cc a code owner soon, we can split if requested.

dan-zheng commented on include/swift/AST/Attr.h:
```diff
@@ -1667,7 +1668,8 @@ class DifferentiableAttr final
   /// The differentiability parameter indices, resolved by the type checker.
-  IndexSubset *ParameterIndices = nullptr;
+  /// The bit stores whether the parameter indices have been computed.
+  llvm::PointerIntPair<IndexSubset *, 1, bool> ParameterIndicesAndBit;
```
`nullptr` is not used because `nullptr` by itself is ambiguous:
* "This attribute has not been type-checked", vs
* "This attribute failing type-checking, no parameter indices could be resolved"

I can add a comment.

dan-zheng replied:
Done.

dan-zheng commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -2020,6 +2025,35 @@ class PatternTypeRequest
+/// - `DerivativeGenericSignature`
+class DifferentiableAttributeTypeCheckRequest
+    : public SimpleRequest<DifferentiableAttributeTypeCheckRequest,
+                           IndexSubset *(DifferentiableAttr *),
```
Currently, `DifferentiableAttributeTypeCheckRequest` takes a `DifferentiableAttr *`and returns an `IndexSubset *` representing parameter indices. It also resolves two `FuncDecl *` (`JVPFunction` and `VJPFunction`) and a `GenericSignature` (`DerivativeGenericSignature`), mutating them in-place in the `DifferentiableAttr *`.

This works fine for now. I thought of two ways to refactor the request to avoid mutation:
* Make `DifferentiableAttributeTypeCheckRequest` return a tuple/struct of the resolved components.
  * This caused many request cycles, since `DifferentiableAttr::get{JVPFunction,VJPFunction,DerivativeGenericSignature}` now all trigger the request. I haven't debugged further.
* Make an individual request for resolving each `@differentiable` attribute component: original `AbstractFuncDecl`(s), parameter indices, JVP/VJP `FuncDecl` (if specified), derivative `GenericSignature` (if specified).
  * I haven't tried this. It seems like a fair amount of work, and the upside is unclear.

dan-zheng replied:
History: https://github.com/apple/swift/pull/28017 moved `@differentiable` attribute type-checking logic from `AttributeChecker::visitDifferentiableAttr` to `DifferentiableAttributeTypeCheckRequest::evaluate`.

The current request approach (return `IndexSubset *` but mutate other components in-place) is the first one that worked.

dan-zheng commented on include/swift/AST/TypeCheckRequests.h:
```diff
@@ -2020,6 +2025,35 @@ class PatternTypeRequest
+class DifferentiableAttributeTypeCheckRequest
+    : public SimpleRequest<DifferentiableAttributeTypeCheckRequest,
+                           IndexSubset *(DifferentiableAttr *),
+                           CacheKind::SeparatelyCached> {
```
Using `CacheKind::Cached` seems preferable so that `DifferentiableAttr` is truly stateless, adhering to the [request evaluator caching vision](https://github.com/apple/swift/blob/tensorflow/docs/RequestEvaluator.md#caching).

I tried changing `CacheKind::SeparatelyCached` to `CacheKind::Cached` (some time ago on `tensorflow` branch), but it did not seem to work: `@differentiable` attributes in non-primary-files were left unchecked.

I'm not sure why `CacheKind::Cached` did not work. I did notice that quite a few other requests (e.g. `InterfaceTypeRequest`) used `CacheKind::SeparatelyCached` for some reason - perhaps that's just for request cycle breaking?

dan-zheng commented:
@CodaFi: would you like to review this PR, or suggest a reviewer? I'm asking you since you seem like a request evaluator owner.

This patch adds a `DifferentiableAttributeTypeCheckRequest` for type-checking `@differentiable` attributes. A request is necessary (instead of just using `AttributeChecker`) to type-check `@differentiable` attributes in non-primary files.

I created two discussion threads regarding possible improvements:
* [Improving `DifferentiableAttributeTypeCheckRequest`, which currently mutates `DifferentiableAttr`.](https://github.com/apple/swift/pull/29231#discussion_r367722559)
* [Using `CacheKind::Cached` for `DifferentiableAttributeTypeCheckRequest` doesn't work.](https://github.com/apple/swift/pull/29231#discussion_r367723033)

---

Any improvement suggestions would be appreciated! If you'd like me to split this patch into two smaller ones ([idea](https://github.com/apple/swift/pull/29231#discussion_r367130563)), I'm happy to do so.

Otherwise, I'll merge this patch within 2 days so that it's in a [development snapshot release](https://github.com/apple/swift/releases) before our `master -> tensorflow` branch merge rotation next week.

dan-zheng commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -530,7 +530,114 @@ swift::matchWitness(
+            req->dumpRef(llvm::dbgs());
+            llvm::dbgs() << '\n';
+          });
+          // FIXME(TF-1014): `@differentiable` attribute diagonstic does not
```
```suggestion
          // FIXME(TF-1014): `@differentiable` attribute diagnostic does not
```

dan-zheng replied:
Done.

dan-zheng committed:
Address review feedback.

- Refactor `DifferentiableAttributeTypeCheckRequest::evaluate` into smaller
  helpers.
  - Consider rewriting helpers as their own requests.
- Add explanatory comment for `DifferentiableAttr::ParameterIndicesAndBit`.
- Refactor `swift::matchWitness` logic for checking `@differentiable`
  attributes into a `matchWitnessDifferentiableAttr` helper.

dan-zheng commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -3401,6 +3399,504 @@ getDerivativeOriginalFunctionType(AnyFunctionType *derivativeFnTy) {
+  original->addDerivativeFunctionConfiguration(
+      {resolvedDiffParamIndices, resultIndices, derivativeGenSig});
+  return resolvedDiffParamIndices;
+}
```
I refactored `DifferentiableAttributeTypeCheckRequest::evaluate` into smaller helper functions. It's only 174 lines long now.

dan-zheng commented:
@swift-ci Please test

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test

dan-zheng commented:
Merging now, happy to address further feedback later!
